File: 1-D DP/Knapsack Variants/Coin Change/solution.py
from typing import List

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [amount + 1] * (amount + 1)
        dp[0] = 0
        
        for i in range(1, amount + 1):
            for coin in coins:
                if i - coin >= 0:
                    dp[i] = min(dp[i], dp[i - coin] + 1)
                    
        return dp[amount] if dp[amount] <= amount else -1

if __name__ == "__main__":
    sol = Solution()
    
    print(f"Test Case 1 (Amt 11): {sol.coinChange([1, 2, 5], 11)}") # 3
    print(f"Test Case 2 (Amt 3): {sol.coinChange([2], 3)}") # -1
    print(f"Test Case 3 (Amt 0): {sol.coinChange([1], 0)}") # 0

File: 1-D DP/Knapsack Variants/Coin Change/solution.java
import java.util.Arrays;

class Solution {
    public int coinChange(int[] coins, int amount) {
        int max = amount + 1;
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, max);
        dp[0] = 0;
        
        for (int i = 1; i <= amount; i++) {
            for (int coin : coins) {
                if (i - coin >= 0) {
                    dp[i] = Math.min(dp[i], dp[i - coin] + 1);
                }
            }
        }
        return (dp[amount] > amount) ? -1 : dp[amount];
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        
        int[] coins1 = {1, 2, 5};
        System.out.println("Test Case 1 (Amt 11): " + sol.coinChange(coins1, 11)); // 3
        
        int[] coins2 = {2};
        System.out.println("Test Case 2 (Amt 3): " + sol.coinChange(coins2, 3)); // -1
        
        int[] coins3 = {1};
        System.out.println("Test Case 3 (Amt 0): " + sol.coinChange(coins3, 0)); // 0
    }
}
